security: rotate certificates

mberhault opened the pull request:
The certificate directory can now be re-read by issuing a SIGHUP signal to the server.

Part of #1674.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14925)
<!-- Reviewable:end -->

tamird commented on pkg/security/certs_test.go:
```diff
@@ -138,6 +140,30 @@ func TestGenerateNodeCerts(t *testing.T) {
+	if err := security.CreateCAPair(
+		certsDir, filepath.Join(certsDir, security.EmbeddedCAKey), 512, time.Hour*48,
+	); err != nil {
+		return errors.Errorf("could not generate CA pair: %v", err)
```
throughout: errors.Wrap

mberhault replied:
why? These are Done.

mberhault replied:
so the goal of this PR was to just add certificate rotation, #14703 is the big one.
I'm happy using this one instead and closing the other one, but the last commit here has the advantage of being self contained.

tschottdorf commented on pkg/base/config.go:
```diff
@@ -220,70 +244,83 @@ func (cfg *Config) PGURL(user *url.Userinfo) (*url.URL, error) {
+
+	cm, err := cfg.GetCertificateManager()
+	if err != nil {
+		return nil
```
uhm, `return err`?

mberhault replied:
probably a good idea. I think it got stuck in my head because barring inconvenient reload, the previous calls will error out if this one does.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (only looking at r7-r9)

---

Reviewed 5 of 6 files at r8, 1 of 1 files at r9.
Review status: 6 of 32 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/security/certificate_manager.go, line 88 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/14925#-Khtr22GgF1qn1TcXycD:-Khtr22GgF1qn1TcXycE:bb27776) ([raw file](https://github.com/cockroachdb/cockroach/blob/48b69ab3afc8e2e688257dac048d338373e01347/pkg/security/certificate_manager.go#L88)):*
> ```Go
> 	// Setup signal handler.
> 	sigs := make(chan os.Signal, 1)
> 	signal.Notify(sigs, syscall.SIGHUP)
> ```

Is `syscall.SIGHUP` defined on windows? This may need to go in a separate file behind a build tag (it's fine to make this a noop on windows)

---

*[pkg/security/gna.stdout, line 0 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/14925#-Khtri1kDgwcbqNqUMGF:-Khtri1kDgwcbqNqUMGG:bg83fly) ([raw file](https://github.com/cockroachdb/cockroach/blob/48b69ab3afc8e2e688257dac048d338373e01347/pkg/security/gna.stdout#L0)):*
This file looks like it wasn't meant to be here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14925#-:-Khtrq1BEtBF6QrX4t52:bukuhbj)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 0 of 31 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/security/certificate_manager.go, line 88 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/14925#-Khtr22GgF1qn1TcXycD:-KhuWBmT1-xBxi6V9r9L:b-ylt3tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/48b69ab3afc8e2e688257dac048d338373e01347/pkg/security/certificate_manager.go#L88)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is `syscall.SIGHUP` defined on windows? This may need to go in a separate file behind a build tag (it's fine to make this a noop on windows)
</blockquote></details>

the unix signals are defined on windows with bogus values, so we don't need to split this up. The test will fail on windows though since it won't trigger a reload.

---

*[pkg/security/gna.stdout, line  at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/14925#-Khtri1kDgwcbqNqUMGF:-KhuWzOoFqXtk4beS8Ti:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/48b69ab3afc8e2e688257dac048d338373e01347/pkg/security/gna.stdout#L)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This file looks like it wasn't meant to be here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14925)*
<!-- Sent from Reviewable.io -->

mberhault commented:
@tamird: I moved the test that uses `signal.Kill` to a separate file with build constraints:
`// +build darwin dragonfly freebsd linux netbsd openbsd solaris`
This is quite a few of them, I got this list from `os/signal/signal_test.go` which seems to do it for the very same reason.

tamird commented:
@mberhault just use `!windows`, there is lots of precedent for this.

mberhault commented:
done.

Commit:
security: implement certs rotation

The certificate directory can now be re-read by issuing a SIGHUP signal
to the server.

Part of #1674.

mberhault commented:
Rebased on top of the merged `--certs-dir` PR. No other changes to this commit.
